Zend/zend_weakrefs.c

```diff
@@ -0,0 +1,242 @@
+/*
+   +----------------------------------------------------------------------+
+   | Zend Engine                                                          |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 1998-2018 Zend Technologies Ltd. (http://www.zend.com) |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: krakjoe@php.net                                             |
+   +----------------------------------------------------------------------+
+*/
+
+/* $Id$ */
+
+#include "zend.h"
+#include "zend_interfaces.h"
+#include "zend_objects_API.h"
+#include "zend_weakrefs.h"
+
+typedef struct _zend_weakref {
+	zval referent;
+	zend_array *root;
+	zend_object std;
+} zend_weakref;
+
+zend_class_entry *zend_ce_weakref;
+zend_object_handlers zend_weakref_handlers;
+
+#define zend_weakref_from(o) ((zend_weakref*)(((char*) o) - XtOffsetOf(zend_weakref, std)))
+#define zend_weakref_fetch(z) zend_weakref_from(Z_OBJ_P(z))
+
+static void zend_weakrefs_dtor(zval *zv) {
+	zend_hash_destroy(Z_PTR_P(zv));
+	FREE_HASHTABLE(Z_PTR_P(zv));
+}
+
+void zend_weakrefs_init(zend_array *root) {
+	zend_hash_init(root, 8, NULL, zend_weakrefs_dtor, 0);
+}
+
+static int zend_weakref_unref(zval *zv) {
+	zend_weakref *wr = (zend_weakref*) Z_PTR_P(zv);
+
+	ZVAL_NULL(&wr->referent);
+
+	wr->root = NULL;
+
+	return ZEND_HASH_APPLY_REMOVE;
+}
+
+static int zend_weakrefs_detach(zval *zv) {
+	zend_hash_apply(Z_PTR_P(zv), zend_weakref_unref);
+
+	return ZEND_HASH_APPLY_REMOVE;
+}
+
+void zend_weakrefs_notify(zend_object *object) {
+	zend_hash_apply(&EG(weakrefs), zend_weakrefs_detach);
```

Comment:
clearly wrong ... my only excuse is it's hot here today ... just ignore me for a while ...